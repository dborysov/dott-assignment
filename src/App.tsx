import React, { FC, useEffect, useRef } from 'react';
import { useCamera } from './hooks/useCamera';
import { usePoseNet } from './hooks/usePoseNet';
import { usePoses } from './hooks/usePoses';
import { useWindowSize } from './hooks/useWindowSize';
import { Button, Eye, Video } from './styles';
import { calculateActualVideoSize } from './utils/calculateActualVideoSize';
import { calculateKeypointPosition } from './utils/calculateKeypointPosition';

const App: FC = () => {
  const screenSize = useWindowSize();
  const { stream, selectedCamera, switchCamera } = useCamera();
  const videoRef = useRef<HTMLVideoElement>(null);
  const net = usePoseNet();
  const { leftEye, rightEye } = usePoses(videoRef, stream, net);

  useEffect(() => {
    if (stream && videoRef.current) {
      // tslint:disable-next-line:no-object-mutation
      videoRef.current.srcObject = stream;
    }
  }, [stream, videoRef.current, selectedCamera]);

  const videoSize = calculateActualVideoSize(
    videoRef.current
      ? { height: videoRef.current.videoHeight, width: videoRef.current.videoWidth }
      : { width: 0, height: 0 },
    screenSize,
  );

  const leftEyePosition = calculateKeypointPosition(screenSize, videoSize, leftEye);
  const rightEyePosition = calculateKeypointPosition(screenSize, videoSize, rightEye);

  return (
    <>
      <Video
        autoPlay={true}
        playsInline={true}
        height={screenSize.height}
        width={screenSize.width}
        ref={videoRef}
      />
      {[leftEyePosition, rightEyePosition].map(
        (eyePosition, i) =>
          eyePosition && (
            <Eye
              key={i}
              style={{
                transform: `translate(${eyePosition.left}px, ${
                  eyePosition.top
                }px) translate(-50%, -50%)`,
              }}
            >
              &#x1F441;
            </Eye>
          ),
      )}
      <Button onClick={switchCamera}>switch camera</Button>
    </>
  );
};

export default App;
