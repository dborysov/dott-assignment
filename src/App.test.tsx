import { render } from 'enzyme';
import React from 'react';
import App from './App';

describe('<App />', () => {
  it('should render as expected', () => {
    expect(render(<App />)).toMatchSnapshot();
  });
});
