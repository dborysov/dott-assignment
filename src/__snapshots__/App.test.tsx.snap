// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render as expected 1`] = `
Array [
  .c0 {
  position: absolute;
  top: 0;
  left: 0;
}

<video
    autoplay=""
    class="c0"
    height="768"
    playsinline=""
    width="1024"
  />,
  .c0 {
  position: absolute;
  right: 1rem;
}

<button
    class="c0"
  >
    switch camera
  </button>,
]
`;
